Guide: Authoring Features - Best Practices (#228)
devcontainers/devcontainers.github.io: _posts/2023-06-14-feature-authoring-best-practices.md

---

> layout: post

Reviewer: Since this is a great guide, we might want to add a link to it from some of our Features docs, i.e. https://containers.dev/implementors/features/ and https://containers.dev/implementors/features-distribution/

title:  "Best Practices: Authoring a Dev Container Feature"
author: "@joshspicer"
authorUrl: https://github.com/joshspicer
---

> Last November I wrote about the basics around [authoring a Dev Container Feature](/guide/author-a-feature). Since then, [hundreds](https://containers.dev/features) of Features have been written by the community. The flexibility of Features has enabled a wide variety of use cases, from installing a single tool to configuring a full development environment.  To that effect, many different patterns for Feature authorship have emerge and the core team has learned a lot about what works well and what doesn't.

Reviewer: ```suggestion
Last November I wrote about the basics around [authoring a Dev Container Feature](/guide/author-a-feature). Since then, [hundreds](https://containers.dev/features) of Features have been written by the community. The flexibility of Features has enabled a wide variety of use cases, from installing a single tool to configuring a full development environment.  To that effect, many different patterns for Feature authorship have emerged, and the core team has learned a lot about what works well and what doesn't.
```

Reviewer: Maybe also add a brief note that the following are recommendations and best practices but not hard requirements?

## Utilize the `test` command

> Bundled with the `devcontainer` cli is the `devcontainer features test` command.  This command is designed to help Feature authors test their Feature in a variety of scenarios.  It is highly recommended that Feature authors use this command to test their Feature before publishing. Some docuemntation on the `test` command can be found [here](https://github.com/devcontainers/cli/blob/main/docs/features/test.md), and an example can be found in the [feature-starter template repo](https://github.com/devcontainers/feature-starter).

Reviewer: nit: Maybe we avoid the word "template" since Templates are a separate thing 😄 

```suggestion
Bundled with the `devcontainer` cli is the `devcontainer features test` command.  This command is designed to help Feature authors test their Feature in a variety of scenarios.  It is highly recommended that Feature authors use this command to test their Feature before publishing. Some docuemntation on the `test` command can be found [here](https://github.com/devcontainers/cli/blob/main/docs/features/test.md), and an example can be found in the [Feature quick start repo](https://github.com/devcontainers/feature-starter).
```

## Feature idempotency

> The most useful Features are idempotent. This means that if a Feature is installed multiple times with different options (something that will become more common with [Feature Dependencies](https://github.com/devcontainers/spec/blob/main/proposals/feature-dependencies.md)), the Feature should be able to handle this gracefully.  This could mean a couple things:

Reviewer: I was thinking this would be a bulleted list after the :

> > There is an open spec proposal for installing the same Feature twice in a given `devcontainer.json`: https://github.com/devcontainers/spec/issues/44.   While the syntax to do so in a given `devcontainer.json` is not yet defined, Feature dependencies will effectively allow for this.

Reviewer: ```suggestion
> There is an open spec proposal for installing the same Feature twice in a given `devcontainer.json`: https://github.com/devcontainers/spec/issues/44.  While the syntax to do so in a given `devcontainer.json` is not yet defined, Feature dependencies will effectively allow for this.
```

### Adding to the $PATH

Features are free to define their own semantics for when two conflicting tools are installed.  Reasonable behavior is to update any references to an installed tool on the PATH to the assets installed from the **last executed Feature**.

> A spec proposal is open for simplifying the process of adding a path to the $PATH variable: https://github.com/devcontainers/spec/issues/251 

For example, if two python Features are installed with different versions, the Feature that was installed last should update the PATH to point to that version of `python` installed.  This ensures that the user is able to use the version of `python` that they selected, and not the version requested by an intermediately installed Feature.

## Writing your install script

> Many of the suggestions in this section may benefit from the 'Feature library'/'code reuse' proposal here https://github.com/devcontainers/spec/blob/main/proposals/features-library.md

> ## Detect Platform/OS

Reviewer: Is this a sub-section of `Writing your install script`?

```suggestion
### Detect Platform/OS
```

> A spec proposal is open for detecting the platform/OS and providing better warnings: https://github.com/devcontainers/spec/issues/58

> Features are often designed to work on subset of possible base images.  For example, the majority of Features in the [`devcontainers/features` repo](https://github.com/devcontainers/features) repo as designed to work broadly with debian-derived images.  The distinction is often simply due to the wide array of base images available, and the fact that many Features will use an OS-specific package manager.

Reviewer: ```suggestion
Features are often designed to work on a subset of possible base images.  For example, the majority of Features in the [`devcontainers/features` repo](https://github.com/devcontainers/features) repo are designed to work broadly with debian-derived images.  The distinction is often simply due to the wide array of base images available, and the fact that many Features will use an OS-specific package manager.
```

One possible way to implement this check is shown below.

```bash
# Source /etc/os-release to get OS info
# Looks something like:
#     PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
#     NAME="Debian GNU/Linux"
#     VERSION_ID="11"
#     VERSION="11 (bullseye)"
#     VERSION_CODENAME=bullseye
#     ID=debian
#     HOME_URL="https://www.debian.org/"
#     SUPPORT_URL="https://www.debian.org/support"
#     BUG_REPORT_URL="https://bugs.debian.org/"
. /etc/os-release
# Store host architecture
architecture="$(dpkg --print-architecture)"

DOCKER_MOBY_ARCHIVE_VERSION_CODENAMES="buster bullseye focal bionic xenial"
if [[ "${DOCKER_MOBY_ARCHIVE_VERSION_CODENAMES}" != *"${VERSION_CODENAME}"* ]]; then
    print_error "Unsupported  distribution version '${VERSION_CODENAME}'. To resolve, either: (1) set feature option '\"moby\": false' , or (2) choose a compatible OS distribution"
    print_error "Supported distributions include:  ${DOCKER_MOBY_ARCHIVE_VERSION_CODENAMES}"
    exit 1
fi

```

> ## Detect the non-root user

Reviewer: Same question as above on this being a sub-section

```suggestion
### Detect the non-root user
```

Feature installation scripts are run as `root`.  In contrast, many dev containers have a `remoteUser` set (either implicitly through [image metadata](https://containers.dev/implementors/spec/#image-metadata) or directly in the `devcontainer.json`).  In a Feature's installation script, one should be mindful of the final user and account for instances where the user is not `root`.

> Feature authors should take advantage of the [`_REMOTE_USER` and similar variables](https://containers.dev/implementors/features/#user-env-var)  injected by conformant implementations.

Reviewer: ```suggestion
Feature authors should take advantage of the [`_REMOTE_USER` and similar variables](https://containers.dev/implementors/features/#user-env-var) injected by conformant implementations.
```

Reviewer: What is a "conformant implementation?" Is it an implementation of the spec? If so, would recommend simplifying the phrasing.

```bash

# _REMOTE_USER is passed in 
USERNAME="${USERNAME:-"${_REMOTE_USER:-"automatic"}"}"

if [ "${USERNAME}" = "auto" ] || [ "${USERNAME}" = "automatic" ]; then
    USERNAME=""
    POSSIBLE_USERS=("vscode" "node" "codespace" "$(awk -v val=1000 -F ":" '$3==val{print $1}' /etc/passwd)")
    for CURRENT_USER in "${POSSIBLE_USERS[@]}"; do
        if id -u ${CURRENT_USER} > /dev/null 2>&1; then
            USERNAME=${CURRENT_USER}
            break
        fi
    done
    if [ "${USERNAME}" = "" ]; then
        USERNAME=root
    fi
elif [ "${USERNAME}" = "none" ] || ! id -u ${USERNAME} > /dev/null 2>&1; then
    USERNAME=root
fi

# ...
# ...

# Install tool and make sure the non-root user has permission to use it
# ...
# chmod -R $USERNAME $TOOL_PATH

```

> ## Implement redundant paths/strategies

Reviewer: Feel free to ignore if this isn't a sub-section!

```suggestion
### Implement redundant paths/strategies
```

> Most Features in [the index today](https://containers.dev/features) have some external/upstream dependency.  Very often these upstream dependencies can change (ie: versioning pattern, rotated GPG key, etc...) that may cause a Feature to fail to install.  To mitigate this, one strategy is to implement multiple paths to install a given tool (if available).  For example, a Feature that installs `go` might try to install it from the upstream package manager, and if not fall back to the a GitHub release.

Reviewer: ```suggestion
Most Features in [the index today](https://containers.dev/features) have some external/upstream dependency.  Very often these upstream dependencies can change (ie: versioning pattern, rotated GPG key, etc...) that may cause a Feature to fail to install.  To mitigate this, one strategy is to implement multiple paths to install a given tool (if available).  For example, a Feature that installs `go` might try to install it from the upstream package manager, and if not fall back to a GitHub release.
```

---
layout: post
title:  "Best Practices: Authoring a Dev Container Feature"
author: "@joshspicer"
authorUrl: https://github.com/joshspicer
---

Last November I wrote about the basics around [authoring a Dev Container Feature](/guide/author-a-feature). Since then, [hundreds](https://containers.dev/features) of Features have been written by the community. The flexibility of Features has enabled a wide variety of use cases, from installing a single tool to setting up specific aspects of a project's development environment that can be shared across repositories.  To that effect, many different patterns for Feature authorship have emerged, and the core team has learned a lot about what works well and what doesn't.

## Utilize the `test` command

> Bundled with the `devcontainer` cli is the `devcontainer features test` command.  This command is designed to help Feature authors test their Feature in a variety of scenarios.  It is highly recommended that Feature authors use this command to test their Feature before publishing. Some documentation on the `test` command can be found [here](https://github.com/devcontainers/cli/blob/main/docs/features/test.md), and an example can be found in the [Feature quick start repo](https://github.com/devcontainers/feature-starter). This repo is updated periodically as new functionality is added to the reference implementation.

Reviewer 2: ```suggestion
Bundled with the `devcontainers/cli` is the `devcontainer features test` command.  This command is designed to help Feature authors test their Feature in a variety of scenarios.  It is highly recommended that Feature authors use this command to test their Feature before publishing. Some documentation on the `test` command can be found [here](https://github.com/devcontainers/cli/blob/main/docs/features/test.md), and an example can be found in the [Feature quick start repo](https://github.com/devcontainers/feature-starter). This repo is updated periodically as new functionality is added to the reference implementation.
```

For Features that install a versioned tool (eg: version x of `go` and version y of `ruby` ), a robust Feature should be able to install multiple versions of the tool.  If your tool has a version manager (java's `SDKMAN`, ruby's `rvm`) it is usually as simple as installing the version manager and then running a command to install the desired version of that tool.

For instances where there isn't an existing version manager available, a well-designed Feature should consider installing distict versions of itself to a well known location.  A pattern that many Features utilize successfully is writing each version of each tool to a central folder and symlinking the "active" version to a folder on the PATH.

Feature can redefine the PATH variable with `containerEnv`, like so:

```bash
# devcontainer-feature.json
"containerEnv": {
    "PATH": "/usr/local/myTool/bin:${PATH}"
}
```

> 🔧 A spec proposal is open for simplifying the process of adding a path to the $PATH variable: https://github.com/devcontainers/spec/issues/251

> To make testing for idempotency easy, [this change to the reference implementation](https://github.com/devcontainers/cli/pull/553) introduces a new mode to the `devcontainer features test` command that will attempt to install a Feature multiple times.  This is useful for testing that a Feature is idempotent, and also for testing that a Feature is able to logically juggle multiple versions of a tool.

Reviewer 2: Clarifying question, currently we do have a way to test installing multiple versions with `devcontainer features test`. However, the CLI/spec doesn't support this in a dev container.

Is this correct? 

> > 🔧 There is an open spec proposal for installing the same Feature twice in a given `devcontainer.json`: https://github.com/devcontainers/spec/issues/44. While the syntax to do so in a given `devcontainer.json` is not yet defined, Feature dependencies will effectively allow for this.

I agree that there's a note here ☝️, however, show we re-iterate to avoid any confusions? maybe like a summary?

Author: Version 0.44.0 of the CLI allows for several instances of a Feature to be installed (via dependencies), so it is possible with the reference implementation today to get into this state

    USERNAME=""
    POSSIBLE_USERS=("vscode" "node" "codespace" "$(awk -v val=1000 -F ":" '$3==val{print $1}' /etc/passwd)")
    for CURRENT_USER in "${POSSIBLE_USERS[@]}"; do
        if id -u ${CURRENT_USER} > /dev/null 2>&1; then
            USERNAME=${CURRENT_USER}
            break
        fi
    done
    if [ "${USERNAME}" = "" ]; then
        USERNAME=root
    fi
elif [ "${USERNAME}" = "none" ] || ! id -u ${USERNAME} > /dev/null 2>&1; then
    USERNAME=root
fi

# Install tool and make sure the non-root user has permission to use it
curl $TOOL_DOWNLOAD_LINK -o $TOOL_PATH
chmod -R $USERNAME $TOOL_PATH

```

### Implement redundant paths/strategies

Most Features in [the index today](https://containers.dev/features) have some external/upstream dependency.  Very often these upstream dependencies can change (ie: versioning pattern, rotated GPG key, etc...) that may cause a Feature to fail to install.  To mitigate this, one strategy is to implement multiple paths to install a given tool (if available).  For example, a Feature that installs `go` might try to install it from the upstream package manager, and if not fall back to a GitHub release.

> Writing several [scenario tests](https://github.com/devcontainers/cli/blob/main/docs/features/test.md#scenarios) that force the Feature to go down distict installation paths will help you catch cases where a given path no longer works.

Reviewer 2: ```suggestion
Writing several [scenario tests](https://github.com/devcontainers/cli/blob/main/docs/features/test.md#scenarios) that force the Feature to go down distinct installation paths will help you catch cases where a given path no longer works.
```

> 🔧 There is an open spec proposal for installing the same Feature twice in a given `devcontainer.json` [(devcontainers/spec#44)](https://github.com/devcontainers/spec/issues/44).  While the syntax to do so in a given `devcontainer.json` is not yet defined, Feature dependencies will effectively allow for this.

For Features that install a versioned tool (eg: version x of `go` and version y of `ruby` ), a robust Feature should be able to install multiple versions of the tool.  If your tool has a version manager (java's `SDKMAN`, ruby's `rvm`) it is usually as simple as installing the version manager and then running a command to install the desired version of that tool.

For instances where there isn't an existing version manager available, a well-designed Feature should consider installing distict versions of itself to a well known location.  A pattern that many Features utilize successfully is writing each version of each tool to a central folder and symlinking the "active" version to a folder on the PATH.

> Feature can redefine the PATH variable with `containerEnv`, like so:

Reviewer 3: nit: "A Feature" or "Features"

One possible way to implement this check is shown below.

```bash
# Source /etc/os-release to get OS info
# Looks something like:
#     PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
#     NAME="Debian GNU/Linux"
#     VERSION_ID="11"
#     VERSION="11 (bullseye)"
#     VERSION_CODENAME=bullseye
#     ID=debian
#     HOME_URL="https://www.debian.org/"
#     SUPPORT_URL="https://www.debian.org/support"
#     BUG_REPORT_URL="https://bugs.debian.org/"
. /etc/os-release
# Store host architecture
architecture="$(dpkg --print-architecture)"

DOCKER_MOBY_ARCHIVE_VERSION_CODENAMES="buster bullseye focal bionic xenial"
if [[ "${DOCKER_MOBY_ARCHIVE_VERSION_CODENAMES}" != *"${VERSION_CODENAME}"* ]]; then
    print_error "Unsupported  distribution version '${VERSION_CODENAME}'. To resolve, either: (1) set feature option '\"moby\": false' , or (2) choose a compatible OS distribution"
    print_error "Supported distributions include:  ${DOCKER_MOBY_ARCHIVE_VERSION_CODENAMES}"
    exit 1
fi
```

> If you're attempting to target a distro that will not have your desired scripting language installed (eg: `bash` is not installed on `alpine`), you may find it useful to use a pattern as shown below.

Reviewer 3: Mention the use of plain `/bin/sh` first, maybe:

If you are targeting distros that may not have your desired scripting language installed (eg: `bash` is often not installed on `alpine` images), you can either use plain `/bin/sh` which is available virtually everywhere or you can check and install the scripting language in a small bootstrap script as shown below.

Author: Thank you, that is worded nicely.

> This example is from the [`common-utils`](https://github.com/devcontainers/features/blob/d934503a050ba84e6b42a006aacd891c4088eb62/src/common-utils/install.sh) Feature and is used to install `bash` on alpine-like distros.

Reviewer 3: nit: With capital A: "Alpine-like" since it's the distro's name. I think I see it all lowercase when authors refer to the image name `alpine`.

      - uses: actions/checkout@v3

      - name: "Install latest devcontainer CLI"
        run: npm install -g @devcontainers/cli
        {% raw %}
      - name: "Generating tests for '${{ matrix.features }}' against '${{ matrix.baseImage }}'"
        run: devcontainer features test  --skip-scenarios -f ${{ matrix.features }} -i ${{ matrix.baseImage }}
        {% endraw %} 
```

### Detect the non-root user

Feature installation scripts are run as `root`.  In contrast, many dev containers have a `remoteUser` set (either implicitly through [image metadata](https://containers.dev/implementors/spec/#image-metadata) or directly in the `devcontainer.json`).  In a Feature's installation script, one should be mindful of the final user and account for instances where the user is not `root`.

Feature authors should take advantage of the [`_REMOTE_USER` and similar variables](https://containers.dev/implementors/features/#user-env-var) injected during the build.

```bash
# _REMOTE_USER is set in the environment during the build to match the effective 'remoteUser'
USERNAME="${USERNAME:-"${_REMOTE_USER:-"automatic"}"}"

# If an existing non-root user matching the POSSIBLE_USERS exists, use it!

> if [ "${USERNAME}" = "auto" ] || [ "${USERNAME}" = "automatic" ]; then

Reviewer 3: I would only show the use of `_REMOTE_USER` without the historical fallbacks. Maybe (untested):
```sh
# Install tool in user's bin folder
mkdir -p "$_REMOTE_USER_HOME/bin"
curl $TOOL_DOWNLOAD_LINK -o "$_REMOTE_USER_HOME/bin/$TOOL"
chown $_REMOTE_USER:$_REMOTE_USER "$_REMOTE_USER_HOME/bin/$TOOL"
chmod 755 "$_REMOTE_USER_HOME/bin/$TOOL"
```
(Using the user's bin folder because it seems a bit contrived to make a regular user the owner of a system-wide binary.)

Author: Good point, we can do everything we need now with those variables. Updating.
